core/src/main/java/io/grpc/ServerImpl.java

```diff
@@ -521,7 +526,13 @@ public void halfClosed() {
           return;
         }
 
-        listener.onHalfClose();
+        GrpcCallContext.enter(callContext);
```

Comment:
Yes, you would get to do it for each listener call, since we may be on a different thread for each one (since the executor decides where to run).